File: frontend-new/src/utils/indicators/nadarayaWatson.ts
/**
 * Nadaraya-Watson Envelope Indicator
 *
 * Implementation based on LuxAlgo's TradingView indicator
 * https://www.tradingview.com/script/Iko0E2kL-Nadaraya-Watson-Envelope-LuxAlgo/
 *
 * Uses Gaussian Kernel Regression to create smooth price envelopes
 * with Mean Absolute Error (MAE) for band calculation.
 */

import { Candle } from './types'

// ============================================
// TYPES
// ============================================

export interface NWEnvelopeParams {
  windowSize: number      // Number of candles (default: 250)
  bandwidth: number       // Gaussian kernel width (default: 8)
  multiplier: number      // MAE multiplier for bands (default: 3)
  source: 'close' | 'hlc3' | 'ohlc4' | 'hl2'
  repaint: boolean        // If true, uses full data (repainting). If false, endpoint method
}

export interface NWEnvelopePoint {
  time: number
  smoothLine: number
  upperBand: number
  lowerBand: number
}

export interface NWSignal {
  time: number
  type: 'buy' | 'sell'
  price: number
}

export interface NWEnvelopeResult {
  points: NWEnvelopePoint[]
  signals: NWSignal[]
  mae: number
}

// ============================================
// DEFAULT PARAMS
// ============================================

export const NW_DEFAULT_PARAMS: NWEnvelopeParams = {
  windowSize: 250,
  bandwidth: 8,
  multiplier: 3,
  source: 'close',
  repaint: true  // Default to repaint mode (like LuxAlgo default)
}

// ============================================
// HELPER FUNCTIONS
// ============================================

/**
 * Gaussian Kernel Function (identical to LuxAlgo)
 * gauss(x, h) => exp(-(x²)/(h²×2))
 */
function gaussianKernel(x: number, h: number): number {
  return Math.exp(-(x * x) / (2 * h * h))
}

/**
 * Extract source price from candle
 */
function getSourcePrice(candle: Candle, source: string): number {
  switch (source) {
    case 'hlc3':
      return (candle.high + candle.low + candle.close) / 3
    case 'ohlc4':
      return (candle.open + candle.high + candle.low + candle.close) / 4
    case 'hl2':
      return (candle.high + candle.low) / 2
    default:
      return candle.close
  }
}

// ============================================
// MAIN CALCULATION FUNCTION
// ============================================

/**
 * Calculate Nadaraya-Watson Envelope
 *
 * Combines LuxAlgo's TradingView logic with performance optimizations.
 *
 * Two modes available:
 * - Repaint Mode: Uses all data points, more accurate visually but repaints
 * - Non-Repaint Mode (Endpoint): Uses only past data, no repainting
 */
export function calculateNadarayaWatsonEnvelope(
  candles: Candle[],
  params: Partial<NWEnvelopeParams> = {}
): NWEnvelopeResult {
  // Merge with defaults
  const {
    windowSize = NW_DEFAULT_PARAMS.windowSize,
    bandwidth = NW_DEFAULT_PARAMS.bandwidth,
    multiplier = NW_DEFAULT_PARAMS.multiplier,
    source = NW_DEFAULT_PARAMS.source,
    repaint = NW_DEFAULT_PARAMS.repaint
  } = params

  const h = bandwidth
  const n = Math.min(candles.length, windowSize)

  // Not enough data
  if (n < 2) {
    return { points: [], signals: [], mae: 0 }
  }

  // Extract source prices and times from most recent candles
  const relevantCandles = candles.slice(-n)
  const src = relevantCandles.map(c => getSourcePrice(c, source))
  const times = relevantCandles.map(c => c.time)

  const points: NWEnvelopePoint[] = []
  const signals: NWSignal[] = []

  if (repaint) {
    // ═══════════════════════════════════════════════════════
    // REPAINT MODE (like LuxAlgo's barstate.islast)
    // More accurate visually, but recalculates on historical bars
    // ═══════════════════════════════════════════════════════

    const smoothLine: number[] = []
    let sae = 0 // Sum of Absolute Errors

    // Step 1: Calculate smoothed line for each point
    // Using double loop like LuxAlgo's Pine Script
    for (let i = 0; i < n; i++) {
      let sum = 0
      let sumW = 0

      for (let j = 0; j < n; j++) {
        // Gaussian weight based on distance between points
        const w = gaussianKernel(i - j, h)
        sum += src[j] * w
        sumW += w
      }

      const y = sum / sumW
      smoothLine.push(y)
      sae += Math.abs(src[i] - y)
    }

    // Step 2: Calculate MAE (Mean Absolute Error) × multiplier
    // Identical to LuxAlgo: sae := sae / math.min(499, n-1) * mult
    const mae = (sae / Math.max(n - 1, 1)) * multiplier

    // Step 3: Build points with bands
    for (let i = 0; i < n; i++) {
      points.push({
        time: times[i],
        smoothLine: smoothLine[i],
        upperBand: smoothLine[i] + mae,
        lowerBand: smoothLine[i] - mae
      })

      // Detect signals (LuxAlgo style)
      // Signal is generated when candle CLOSES above/below the band (not just touches)
      // - Candle CLOSES ABOVE upper band = SELL signal (overbought, expect reversal down)
      // - Candle CLOSES BELOW lower band = BUY signal (oversold, expect reversal up)

      const currClose = src[i]

      // Use current candle's band values
      const upper = smoothLine[i] + mae
      const lower = smoothLine[i] - mae

      // SELL signal: candle CLOSES ABOVE upper band (overbought)
      if (currClose > upper) {
        signals.push({
          time: times[i],
          type: 'sell',
          price: currClose
        })
      }

      // BUY signal: candle CLOSES BELOW lower band (oversold)
      if (currClose < lower) {
        signals.push({
          time: times[i],
          type: 'buy',
          price: currClose
        })
      }
    }

    console.log(`🔍 NW Repaint Mode: Generated ${signals.length} signals from ${n} candles`)
    if (signals.length > 0) {
      console.log('📍 Signals:', signals.slice(0, 5), '...')
    }
    return { points, signals, mae }

  } else {
    // ═══════════════════════════════════════════════════════
    // NON-REPAINT MODE / ENDPOINT (like LuxAlgo's non-repaint option)
    // More reliable for real trading, uses only past data
    // ═══════════════════════════════════════════════════════

    // Pre-calculate coefficients (LuxAlgo optimization)
    // This avoids recalculating weights for each point
    const coefs: number[] = []
    let den = 0

    for (let i = 0; i < n; i++) {
      const w = gaussianKernel(i, h)
      coefs.push(w)
      den += w
    }

    // Calculate smoothed line for each point
    // Using endpoint method (only looks backward)
    const smoothLine: number[] = []

    for (let i = 0; i < n; i++) {
      let out = 0
      let localDen = 0

      // Only use data up to current index (non-repaint)
      const lookback = Math.min(i + 1, n)
      for (let j = 0; j < lookback; j++) {
        const w = coefs[j]
        const idx = i - j
        if (idx >= 0) {
          out += src[idx] * w
          localDen += w
        }
      }

      smoothLine.push(localDen > 0 ? out / localDen : src[i])
    }

    // Calculate MAE using SMA-style (like LuxAlgo's ta.sma)
    // mae = ta.sma(math.abs(src - out), 499) * mult
    const maeWindow = Math.min(n, 100) // Smaller window for performance

    for (let i = 0; i < n; i++) {
      // Calculate local MAE for this point
      let maeSum = 0
      let maeCount = 0

      const start = Math.max(0, i - maeWindow + 1)
      for (let k = start; k <= i; k++) {
        maeSum += Math.abs(src[k] - smoothLine[k])
        maeCount++
      }

      const localMae = maeCount > 0 ? (maeSum / maeCount) * multiplier : 0

      points.push({
        time: times[i],
        smoothLine: smoothLine[i],
        upperBand: smoothLine[i] + localMae,
        lowerBand: smoothLine[i] - localMae
      })

      // Detect signals (LuxAlgo style)
      // Signal is generated when candle CLOSES above/below the band
      const currClose = src[i]
      const upper = points[i].upperBand
      const lower = points[i].lowerBand

      // SELL signal: candle CLOSES ABOVE upper band (overbought)
      if (currClose > upper) {
        signals.push({
          time: times[i],
          type: 'sell',
          price: currClose
        })
      }

      // BUY signal: candle CLOSES BELOW lower band (oversold)
      if (currClose < lower) {
        signals.push({
          time: times[i],
          type: 'buy',
          price: currClose
        })
      }
    }

    const finalMae = points.length > 0
      ? points[points.length - 1].upperBand - points[points.length - 1].smoothLine
      : 0

    console.log(`🔍 NW Non-Repaint Mode: Generated ${signals.length} signals from ${n} candles`)
    if (signals.length > 0) {
      console.log('📍 Signals:', signals.slice(0, 5), '...')
    }
    return { points, signals, mae: finalMae }
  }
}

// ============================================
// INDICATOR CONFIG (for IndicatorEngine integration)
// ============================================

export interface NWEnvelopeConfig {
  id: string
  type: 'NWENVELOPE'
  enabled: boolean
  displayType: 'overlay'
  color: string
  lineWidth: number
  params: {
    windowSize: number
    bandwidth: number
    multiplier: number
    source: 'close' | 'hlc3' | 'ohlc4' | 'hl2'
    repaint: boolean
  }
}

export const NW_ENVELOPE_PRESET: Partial<NWEnvelopeConfig> = {
  displayType: 'overlay',
  color: '#2962FF',  // Blue for smooth line
  lineWidth: 2,
  params: {
    windowSize: 250,
    bandwidth: 8,
    multiplier: 3,
    source: 'close',
    repaint: true
  }
}

// ============================================
// COLORS (LuxAlgo style)
// ============================================

export const NW_COLORS = {
  smoothLine: '#2962FF',              // Blue
  upperBand: 'rgba(239, 83, 80, 0.8)', // Red (dnCss)
  lowerBand: 'rgba(38, 166, 154, 0.8)', // Teal (upCss)
  buySignal: '#26A69A',                // Teal
  sellSignal: '#EF5350',               // Red
  fillArea: 'rgba(41, 98, 255, 0.05)'  // Light blue fill
}

File: apps/api-python/infrastructure/indicators/nadaraya_watson.py
"""Nadaraya-Watson Envelope Calculator

Implementation of Gaussian Kernel Regression for price smoothing
with upper and lower bands based on ATR or standard deviation.

Based on the TradingView Nadaraya-Watson Envelope indicator.
"""

import math
from decimal import Decimal
from typing import Any, Dict, List, Optional

from .base import BaseIndicatorCalculator, Candle, IndicatorResult


class NadarayaWatsonCalculator(BaseIndicatorCalculator):
    """
    Nadaraya-Watson Envelope Calculator

    Uses Gaussian Kernel Regression to create a smoothed price line
    with upper and lower bands.

    Parameters:
        bandwidth (int): Lookback period for kernel regression (default: 8)
        mult (float): Multiplier for band width (default: 3.0)
        src (str): Source price - 'close', 'hl2', 'hlc3', 'ohlc4' (default: 'close')
        use_atr (bool): Use ATR for bands instead of kernel deviation (default: True)
        atr_period (int): ATR period if use_atr is True (default: 14)

    Output values:
        - value: The smoothed Nadaraya-Watson line
        - upper: Upper band
        - lower: Lower band
        - signal: 1 (bullish), -1 (bearish), 0 (neutral)
    """

    name = "nadaraya_watson"
    required_candles = 50

    def __init__(self, parameters: Optional[Dict[str, Any]] = None):
        super().__init__(parameters)

        self.bandwidth = int(self.get_parameter("bandwidth", 8))
        self.mult = float(self.get_parameter("mult", 3.0))
        self.src = self.get_parameter("src", "close")
        self.use_atr = self.get_parameter("use_atr", True)
        self.atr_period = int(self.get_parameter("atr_period", 14))

    def _validate_parameters(self) -> None:
        """Validate parameters"""
        if self.get_parameter("bandwidth") is not None:
            bandwidth = int(self.get_parameter("bandwidth"))
            if bandwidth < 1 or bandwidth > 100:
                raise ValueError("bandwidth must be between 1 and 100")

        if self.get_parameter("mult") is not None:
            mult = float(self.get_parameter("mult"))
            if mult <= 0:
                raise ValueError("mult must be positive")

    def _get_source_price(self, candle: Candle) -> float:
        """Get the source price from a candle"""
        if self.src == "close":
            return float(candle.close)
        elif self.src == "hl2":
            return (float(candle.high) + float(candle.low)) / 2
        elif self.src == "hlc3":
            return (float(candle.high) + float(candle.low) + float(candle.close)) / 3
        elif self.src == "ohlc4":
            return (float(candle.open) + float(candle.high) + float(candle.low) + float(candle.close)) / 4
        else:
            return float(candle.close)

    def _gaussian_kernel(self, x: float, bandwidth: float) -> float:
        """Calculate Gaussian kernel weight"""
        return math.exp(-0.5 * (x / bandwidth) ** 2)

    def _calculate_atr(self, candles: List[Candle], period: int) -> float:
        """Calculate Average True Range"""
        if len(candles) < period + 1:
            return 0.0

        true_ranges = []
        for i in range(1, len(candles)):
            high = float(candles[i].high)
            low = float(candles[i].low)
            prev_close = float(candles[i-1].close)

            tr = max(
                high - low,
                abs(high - prev_close),
                abs(low - prev_close)
            )
            true_ranges.append(tr)

        if len(true_ranges) < period:
            return sum(true_ranges) / len(true_ranges) if true_ranges else 0.0

        return sum(true_ranges[-period:]) / period

    def _calculate_kernel_regression(self, prices: List[float]) -> List[float]:
        """Calculate Nadaraya-Watson kernel regression values"""
        n = len(prices)
        nw_values = []

        for i in range(n):
            weighted_sum = 0.0
            weight_sum = 0.0

            for j in range(n):
                weight = self._gaussian_kernel(i - j, self.bandwidth)
                weighted_sum += weight * prices[j]
                weight_sum += weight

            if weight_sum > 0:
                nw_values.append(weighted_sum / weight_sum)
            else:
                nw_values.append(prices[i])

        return nw_values

    def calculate(self, candles: List[Candle]) -> IndicatorResult:
        """Calculate Nadaraya-Watson Envelope values

        Args:
            candles: List of OHLCV candles (oldest first)

        Returns:
            IndicatorResult with 'value', 'upper', 'lower', 'signal'
        """
        if len(candles) < self.required_candles:
            raise ValueError(f"Need at least {self.required_candles} candles, got {len(candles)}")

        prices = [self._get_source_price(c) for c in candles]

        nw_values = self._calculate_kernel_regression(prices)

        current_nw = nw_values[-1]

        if self.use_atr:
            band_width = self._calculate_atr(candles, self.atr_period) * self.mult
        else:
            deviations = [abs(prices[i] - nw_values[i]) for i in range(len(prices))]
            std_dev = (sum(d**2 for d in deviations) / len(deviations)) ** 0.5
            band_width = std_dev * self.mult

        upper = current_nw + band_width
        lower = current_nw - band_width

        current_close = float(candles[-1].close)
        if current_close > upper:
            signal = -1  # Bearish - price above upper band
        elif current_close < lower:
            signal = 1   # Bullish - price below lower band
        else:
            signal = 0   # Neutral

        return IndicatorResult(
            name=self.name,
            timestamp=candles[-1].timestamp,
            values={
                "value": Decimal(str(round(current_nw, 8))),
                "upper": Decimal(str(round(upper, 8))),
                "lower": Decimal(str(round(lower, 8))),
                "signal": Decimal(str(signal)),
                "band_width": Decimal(str(round(band_width, 8)))
            }
        )
